apis/apiextensions/fn/v1alpha1/functionio_types.go

```diff
@@ -0,0 +1,266 @@
+	// pipeline. Functions may mutate any part of the desired state they are
+	// concerned with, and must pass through any part of the desired state that
+	// they are not concerned with. Functions may omit desired state that they
+	// are unconcerned with as long as they don't need to pass it through. For
+	// example if desired.composite is unset when the function is called it does
+	// not need to set it. If desired.composite is set the function may mutate
+	// it and must return it.
+	// +optional
+	Desired Desired `json:"desired,omitempty"`
+
+	// Items is a list of Crossplane resources - either XRs or MRs.
+	//
+	// A function will read this field in the input FunctionIO and populate
+	// this field in the output FunctionIO.
+	// +kubebuilder:validation:EmbeddedResource
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Items []runtime.RawExtension `json:"items"`
+
+	// Results is an optional list that can be used by function to emit results
+	// for observability and debugging purposes.
+	// +optional
+	Results []Result `json:"results,omitempty"`
+}
+
+// Observed state at the beginning of a function pipeline invocation.
+type Observed struct {
+	// Composite reflects the observed state of the XR this function reconciles.
+	Composite ObservedComposite `json:"composite"`
+
+	// Resources reflect the observed state of any extant composed resources
+	// this function reconciles. Only composed resources that currently exist in
+	// the API server (i.e. have been created and not yet deleted) are included.
+	// +optional
+	Resources []ObservedResource `json:"resources,omitempty"`
+}
+
+// An ObservedComposite resource.
+type ObservedComposite struct {
+	// Resource reflects the observed XR.
+	Resource runtime.RawExtension `json:"resource"`
+
+	// ConnectionDetails reflects the observed connection details of the XR.
+	ConnectionDetails []ExplicitConnectionDetail `json:"connectionDetails"`
+}
+
+// An ObservedResource represents an observed composed resource.
+type ObservedResource struct {
+	// Name of the observed resource. Must be unique within the array of
+	// observed resources. Corresponds to the name entry in a Composition's
+	// resources array, and the name entry in the desired resources array.
+	Name string `json:"name"`
+
+	// Resource reflects the observed composed resource.
+	Resource runtime.RawExtension `json:"resource"`
+
+	// ConnectionDetails reflects the observed connection details of the
+	// composed resource.
+	ConnectionDetails []ExplicitConnectionDetail `json:"connectionDetails"`
+}
+
+// Desired state of a function pipeline invocation.
+type Desired struct {
+	// Composite reflects the desired state of the XR this function reconciles.
+	// +optional
+	Composite DesiredComposite `json:"composite,omitempty"`
+
+	// Resources reflect the desired state of composed resources, including
+	// those that do not yet exist.
+	// +optional
+	Resources []DesiredResource `json:"resources,omitempty"`
```

Comment:
I'd prefer to stick to Kubernetes API conventions here. While this isn't a "real" custom resource, it is a KRM-like object (at least it has an `apiVersion` and `kind`). I know maps are often easier to work with, but we're pretty committed to lists of named subjects everywhere in the Crossplane APIs so I think it will be less surprising to be consistent.

https://github.com/kubernetes/community/blob/0b036b/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps